libglusterfs/src/event-epoll.c

```diff
@@ -70,60 +70,60 @@ static int
 __event_slot_alloc(struct event_pool *event_pool, int fd,
                    int notify_poller_death, struct event_slot_epoll **slot)
 {
-    int i = 0;
     int j = 0;
-    int table_idx;
+    int table_idx = 0;
     int gen;
     struct event_slot_epoll *table = NULL;
 
 retry:
 
-    while (i < EVENT_EPOLL_TABLES) {
-        table = event_pool->ereg[i];
+    while (table_idx < EVENT_EPOLL_TABLES) {
+        table = event_pool->ereg[table_idx];
         if (table) {
-            if (table->slots_used == EVENT_EPOLL_SLOTS)
-                return -1;
-            else
+            if (table->slots_used == EVENT_EPOLL_SLOTS) {
+                table_idx++;
+                continue;
+            } else {
                 break; /* break out of the loop */
+            }
         } else {
-            table = __event_newtable(event_pool, i);
-            if (!table)
+            table = __event_newtable(event_pool, table_idx);
+            if (!table) {
                 return -1;
-            else
-                break; /* break out of the loop */
+            } else {
+                break;
+            }
         }
-
-        i++;
     }
 
-    table_idx = i;
-
-    for (j = 0; j < EVENT_EPOLL_SLOTS; j++) {
-        if (table[j].fd == -1) {
-            /* wipe everything except bump the generation */
-            gen = table[j].gen;
-            memset(&table[j], 0, sizeof(table[j]));
-            table[j].fd = fd;
-            table[j].gen = gen + 1;
+    if (table_idx < EVENT_EPOLL_TABLES) {
```

Comment:
This is not correct. If `table_idx` is equal to `EVENT_EPOLL_TABLES`, which means that there are no tables available, `j` will be 0 and the check in line 124 will incorrectly assume that a slot has been found.

I think it would be better to just check the error condition (i.e. `table_idx >= EVENT_EPOLL_TABLES`) and return -1 in that case. Otherwise continue normally.

Reply:
Done.